### PR TITLE
Add projects stat fields

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This adds some stat fields to the project model in Tenant, and adds a method to update them asynchronously. Eventually we want to have the updates populated via Ensign so the method is more of a stop gap measure for now.

Fixes SC-16445

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Describe how reviewers can test this change to be sure that it works correctly. Add a checklist if possible.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [ ]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?

### CODE REVIEW

File: pkg/tenant/apikeys.go [hunk 1]
```diff
@@ -225,6 +227,11 @@ func (s *Server) ProjectAPIKeyCreate(c *gin.Context) {
 		Modified:     key.Modified.Format(time.RFC3339Nano),
 	}
 
+	// Update project stats in the background
+	s.tasks.QueueContext(sentry.CloneContext(c), tasks.TaskFunc(func(ctx context.Context) error {
+		return s.UpdateProjectStats(ctx, key.ProjectID)
+	}), tasks.WithError(fmt.Errorf("could not update stats for project %s", key.ProjectID.String())))
```
Comment (pdeziel) [inline pkg/tenant/apikeys.go:233]: Only doing this in `ProjectAPIKeyCreate` and `ProjectTopicCreate` for right now so the status shows up properly in the frontend. I'm hoping that we can replace with the ensign topic once Tenant is setup for that.

File: pkg/tenant/projects.go [hunk 2]
```diff
@@ -514,3 +514,67 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
...
+			return err
+		}
+
+		numAPIKeys += uint64(len(reply.APIKeys))
+		if reply.NextPageToken == "" {
+			break
+		}
+
+		req.NextPageToken = reply.NextPageToken
+	}
```
Comment (bbengfort) [inline pkg/tenant/projects.go:547]: Yeh, I can see the issue; this isn't particularly performant -- probably what we should do is have a Quarterdeck endpoint that returns the number of API keys -- I'll create a story for that.

File: pkg/tenant/projects.go [hunk 3]
```diff
@@ -514,3 +514,67 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
...
+
+	// Count the number of topics in trtl.
+	var next *pg.Cursor
+	for {
+		var topics []*db.Topic
+		if topics, next, err = db.ListTopics(ctx, projectID, next); err != nil {
+			return err
+		}
+
+		numTopics += uint64(len(topics))
```
[the remark this answers is not in the document]
Comment (bbengfort) [reply pkg/tenant/projects.go:557]: We probably would want to differentiate them in the tenant stats. We probably also want an Ensign endpoint for this as well.

File: pkg/tenant/projects.go [hunk 4]
```diff
@@ -514,3 +514,67 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
...
+	}
+
+	// Retrieve the project from trtl.
+	var project *db.Project
+	if project, err = db.RetrieveProject(ctx, projectID); err != nil {
+		return err
+	}
+
+	// TODO: A project write in between here could cause updates to be stomped (e.g. if
+	// the project name is updated it will be overwritten here).
```
Comment (bbengfort) [inline pkg/tenant/projects.go:570]: This is a good point and there needs to be a transaction read then write in trtl to allow for a read then a write.

**Review by bbengfort — COMMENTED** [review]
This does make sense to me - though I think we need to streamline and parallelize the Quarterdeck and Ensign queries. Let me see if I can quickly create endpoints for project info -- then in the meantime you can perform the data fetching in a go routine?

File: cmd/debug/main.go [hunk 5]
```diff
@@ -208,7 +208,7 @@ func generate(c *cli.Context) (err error) {
 				}
 
 				msg++
-				publisher.Publish(&api.Event{
+				publisher.Publish("generator", &api.Event{
```
Comment (pdeziel) [inline cmd/debug/main.go:211]: Can you verify this is the correct usage? I didn't run the script to test it.

File: pkg/tenant/projects.go [hunk 6]
```diff
@@ -514,3 +517,96 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
...
+	rpc := func(fn func(context.Context) (uint64, error), i int, res *uint64) {
+		defer wg.Done()
+		*res, errs[i] = fn(ctx)
+	}
+	wg.Add(2)
+	go rpc(countAPIKeys, 0, numAPIKeys)
+	go rpc(countTopics, 1, numTopics)
+	wg.Wait()
+
+	// Return if both RPCs errored.
```
Comment (pdeziel) [inline pkg/tenant/projects.go:583]: Should we return if only one of them errors to keep the Tenant view more consistent?

File: cmd/debug/main.go (hunk 5 above)
Comment (bbengfort) [reply cmd/debug/main.go:211]: I think that needs to be a topic id - but don't worry about it; we'll need to add authentication etc. Adding "generator" here ensures that ensign builds which is enough for now.

File: pkg/tenant/projects.go [hunk 7]
```diff
@@ -514,3 +517,96 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
...
+	errs := make([]error, 2)
+	res := make([]uint64, 2)
+	rpc := func(fn func(context.Context) (uint64, error), i int) {
+		defer wg.Done()
+		res[i], errs[i] = fn(ctx)
+	}
+	wg.Add(2)
+	go rpc(countAPIKeys, 0)
+	go rpc(countTopics, 1)
+	wg.Wait()
```
Comment (bbengfort) [inline pkg/tenant/projects.go:581]: Nicely concise!

File: pkg/tenant/projects.go (hunk 6 above)
Comment (bbengfort) [reply pkg/tenant/projects.go:583]: Yes, you're probably right; e.g. if the user doesn't have read:topics we still want to know how many api keys.

**Review by bbengfort — APPROVED** [review]
Looks good - glad this is a bit easier than listing everything from Ensign and Quarterdeck!

**Pull request merged**